deps/thpool/thpool.c

```diff
@@ -435,38 +437,29 @@ static void* thread_do(struct thread* thread_p) {
 
   while (thpool_p->state & (THPOOL_KEEP_ALIVE | THPOOL_TERMINATE_WHEN_EMPTY)) {
 
-    bsem_wait(thpool_p->jobqueue.has_jobs);
     LOG_IF_EXISTS("debug", "Thread-%d is running iteration", thread_p->id)
-    if (thpool_p->state & (THPOOL_KEEP_ALIVE | THPOOL_TERMINATE_WHEN_EMPTY)) {
-
-      pthread_mutex_lock(&thpool_p->thcount_lock);
-      size_t iter_id = thpool_p->num_threads_working++;
-      pthread_mutex_unlock(&thpool_p->thcount_lock);
-
-      /* Read job from queue and execute it */
-      void (*func_buff)(void*);
-      void* arg_buff;
-      job* job_p = priority_queue_pull(&thpool_p->jobqueue, iter_id);
-      if (job_p) {
-        func_buff = job_p->function;
-        arg_buff = job_p->arg;
-        func_buff(arg_buff);
-        rm_free(job_p);
-      }
 
-      pthread_mutex_lock(&thpool_p->thcount_lock);
-      thpool_p->num_threads_working--;
-      if (job_p) thpool_p->total_jobs_done++;
-      if (thpool_p->num_threads_working == 0) {
-	      LOG_IF_EXISTS("debug", "All threads are idle")
-	      pthread_cond_signal(&thpool_p->threads_all_idle);
-      }
-      if (thpool_p->state == THPOOL_TERMINATE_WHEN_EMPTY && priority_queue_len(&thpool_p->jobqueue) == 0) {
-        LOG_IF_EXISTS("verbose", "Job queue is empty - terminating thread %d", thread_p->id);
-        thpool_p->state = THPOOL_UNINITIALIZED;
-      }
-      pthread_mutex_unlock(&thpool_p->thcount_lock);
+
+    /* Read job from queue and execute it */
+    void (*func_buff)(void*);
+    void* arg_buff;
+    job* job_p = priority_queue_pull(&thpool_p->jobqueue);
+    if (job_p) {
+      func_buff = job_p->function;
+      arg_buff = job_p->arg;
+      func_buff(arg_buff);
+      rm_free(job_p);
+    }
+
+    pthread_mutex_lock(&thpool_p->thcount_lock);
+    if (job_p) thpool_p->total_jobs_done++;
+    if (thpool_p->state == THPOOL_TERMINATE_WHEN_EMPTY && priority_queue_len(&thpool_p->jobqueue) == 0) {
+      LOG_IF_EXISTS("verbose", "Job queue is empty - terminating thread %d", thread_p->id);
+      thpool_p->state = THPOOL_UNINITIALIZED;
+      thpool_p->jobqueue.should_run = false;
+      pthread_cond_broadcast(&thpool_p->jobqueue.has_jobs);
```

Comment:
Use the queue lock instead, see if we can remove the thcount_lock, and then try to optimize (+check performance)